models/user_setting.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"strings"
+
+	"xorm.io/builder"
+)
+
+// UserSetting is a key value store of user settings
+type UserSetting struct {
+	ID     int64  `xorm:"pk autoincr"`
+	UserID int64  `xorm:"index"`              // to load all of someone's settings
+	Key    string `xorm:"varchar(255) index"` // ensure key is always lowercase
+	Value  string `xorm:"text"`
+}
+
+// BeforeInsert will be invoked by XORM before inserting a record
+func (setting *UserSetting) BeforeInsert() {
+	setting.Key = strings.ToLower(setting.Key)
+}
+
+// GetUserSetting returns specific settings from user
+func GetUserSetting(uid int64, keys []string) ([]*UserSetting, error) {
+	settings := make([]*UserSetting, 0, 5)
+	if err := x.
+		Where("user_id=?", uid).
+		And(builder.In("key", keys)).
+		Asc("id").
+		Find(&settings); err != nil {
+		return nil, err
+	}
+	return settings, nil
+}
+
+// GetUserAllSettings returns all settings from user
+func GetUserAllSettings(uid int64) ([]*UserSetting, error) {
+	settings := make([]*UserSetting, 0, 5)
+	if err := x.
+		Where("user_id=?", uid).
+		Asc("id").
+		Find(&settings); err != nil {
+		return nil, err
+	}
+	return settings, nil
+}
+
+// AddUserSetting adds a specific setting for a user
+func AddUserSetting(setting *UserSetting) error {
```

Comment:
We need a `insertOrUpdateUserSetting` method to do the thing.

Reply:
maybe `Set()`?

Reply:
Both the name are OK for me.

Reply:
updated